File: content/exercises/economics-data-science/time-series/rolling-stats/checker.py
import io
import runpy
import sys


def main() -> None:
    submission_path = sys.argv[1]
    output_path = sys.argv[2]
    tests: list[tuple[str, bool, str]] = []
    feedback: list[str] = []
    status = "failed"

    try:
        namespace = runpy.run_path(submission_path, run_name="__main__")
        tests.append(("code runs", True, "Code executed without errors."))
    except Exception as error:  # noqa: BLE001
        namespace = {}
        tests.append(("code runs", False, "Code did not run successfully."))
        feedback.append(f"Execution error: {error}")

    df = namespace.get("df")
    if df is None:
        tests.append(("df exists", False, "Define DataFrame variable df."))
    else:
        has_ma3 = "ma3" in getattr(df, "columns", [])
        has_ma12 = "ma12" in getattr(df, "columns", [])
        tests.append(("ma3 column", has_ma3, "Create df['ma3'] with rolling window 3."))
        tests.append(("ma12 column", has_ma12, "Create df['ma12'] with rolling window 12."))

    passed_all = tests and all(item[1] for item in tests)
    if passed_all:
        status = "passed"
        feedback.append("Correct solution submitted.")
    elif not feedback:
        feedback.append("Compute rolling means for ma3 and ma12 on df['gdp'].")

    output_stream = io.StringIO()
    output_stream.write(f"STATUS:{status}\n")
    for item in feedback:
        output_stream.write(f"FEEDBACK:{item}\n")
    for name, passed, message in tests:
        output_stream.write(f"TEST:{name}|{'pass' if passed else 'fail'}|{message}\n")
    with open(output_path, "w", encoding="utf-8") as handle:
        handle.write(output_stream.getvalue())


if __name__ == "__main__":
    main()

File: content/exercises/intermediate-r/tidyr-purrr/pivot-wider/checker.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 2) stop("Usage: checker.R <submission_path> <output_path>")
submission_path <- args[[1]]
output_path <- args[[2]]

test_results <- list()
feedback <- c()
status <- "failed"
record_test <- function(name, passed, message) {
  test_results[[length(test_results) + 1]] <<- list(name = name, passed = passed, message = message)
}

env <- new.env(parent = baseenv())
execution_ok <- TRUE
execution_error <- NULL
tryCatch({
  source(submission_path, local = env)
}, error = function(e) {
  execution_ok <<- FALSE
  execution_error <<- conditionMessage(e)
})

if (!execution_ok) {
  record_test("code runs", FALSE, "Your code did not run successfully.")
  feedback <- c(feedback, paste("Execution error:", execution_error))
} else {
  record_test("code runs", TRUE, "Code executed without errors.")
  has_wide <- exists("gdp_wide", envir = env, inherits = FALSE)
  record_test("gdp_wide exists", has_wide, "Define widened table as gdp_wide.")
  if (has_wide) {
    gdp_wide <- get("gdp_wide", envir = env, inherits = FALSE)
    is_df <- is.data.frame(gdp_wide)
    has_cols <- all(c("country", "2022", "2023") %in% colnames(gdp_wide))
    record_test("gdp_wide data frame", is_df, "gdp_wide should be a data frame.")
    record_test("gdp_wide columns", has_cols, "Use pivot_wider to create 2022 and 2023 columns.")
  }
}

passed_all <- length(test_results) > 0 && all(vapply(test_results, function(t) isTRUE(t$passed), logical(1)))
if (passed_all) {
  status <- "passed"; feedback <- c(feedback, "Correct solution submitted.")
} else if (length(feedback) == 0) {
  feedback <- c(feedback, "Create gdp_wide using pivot_wider(names_from = year, values_from = gdp).")
}

output_lines <- c(paste0("STATUS:", status))
for (msg in feedback) output_lines <- c(output_lines, paste0("FEEDBACK:", msg))
for (t in test_results) output_lines <- c(output_lines, paste0("TEST:", t$name, "|", if (isTRUE(t$passed)) "pass" else "fail", "|", t$message))
writeLines(output_lines, con = output_path)
